### PR TITLE
Stage location type [WIP]

### PR DESCRIPTION
**Pull request by twicki**
## Technical Description

Add a location type to the stage. This allows for code-generation that loops over edges/vertices/triangles as the "domain-loop"

### Resolves / Enhances

Resolves GT-80

### Example

See example in dawn-c unittest

### Dependencies

This PR is built on top of https://github.com/MeteoSwiss-APN/dawn/pull/316, merge that one before reviewing this

### CODE REVIEW

File: dawn/src/dawn/AST/ASTExpr.h [hunk 1]
```diff
@@ -477,7 +484,7 @@ class FieldAccessExpr : public Expr {
   Array3i argumentOffset_;
 
   // Negate the offset (this allows writing `in(-off)`)
-  bool negateOffset_;
+  bool negateOffset_;
```
Comment (mroethlin) [inline dawn/src/dawn/AST/ASTExpr.h]: clang format?

File: dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp [hunk 2]
```diff
@@ -69,7 +69,21 @@ void ASTStencilBody::visit(const std::shared_ptr<iir::ReturnStmt>& stmt) {
   ss_ << ";\n";
 }
 void ASTStencilBody::visit(const std::shared_ptr<iir::ReductionOverNeighborExpr>& expr) {
-  ss_ << std::string(indent_, ' ') << "reduce(cellNeighboursOfCell(m_mesh, t), ";
+  std::string typeString;
+  switch(expr->getRhsLocation()) {
```
Comment (mroethlin) [inline dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp]: wrap into lambdas?

File: dawn/src/dawn/CodeGen/CXXNaive-ico/CXXNaiveCodeGen.cpp [hunk 3]
```diff
@@ -136,7 +136,24 @@ void CXXNaiveIcoCodeGen::generateStencilWrapperCtr(
 
   std::string ctrArgs("(dom");
   for(auto APIfieldID : APIFields) {
-    StencilWrapperConstructor.addArg("Field<double>& " +
+
+    std::string typeString = "";
+
+    switch(metadata.getLocationTypeFromAccessID(APIfieldID)) {
```
Comment (mroethlin) [inline dawn/src/dawn/CodeGen/CXXNaive-ico/CXXNaiveCodeGen.cpp]: wrap into lambdas?

File: dawn/src/dawn/IIR/StencilMetaInformation.cpp [hunk 4]
```diff
@@ -426,6 +426,20 @@ int StencilMetaInformation::addTmpField(FieldAccessType type, const std::string&
...
+  return FieldAccessIdToLocationType_.count(AccessID) != 0;
+}
+
+ast::Expr::LocationType StencilMetaInformation::getLocationTypeFromAccessID(int AccessID) const {
+  DAWN_ASSERT(getIsUnorderedFromAcessID(AccessID));
+  return FieldAccessIdToLocationType_.at(AccessID);
+}
+
+void StencilMetaInformation::addAccessIDLocationPair(int AccessID,
+                                                     ast::Expr::LocationType location) {
```
Comment (mroethlin) [inline dawn/src/dawn/IIR/StencilMetaInformation.cpp]: assert that not already present

File: dawn/src/dawn/AST/ASTExpr.h (hunk 1 above)
Comment (mroethlin) [inline dawn/src/dawn/AST/ASTExpr.h]: clang format?

File: dawn/prototype/driver.cpp [hunk 5]
```diff
@@ -6,22 +6,20 @@
...
-  Field<double> in(m), out(m);
+  CellField<double> out(m);
+  EdgeField<double> in(m);
 
-  for(auto& f : m.faces()) {
-    auto center_x = w / 2.f - (1.f / 3) * (f.vertex(0).x() + f.vertex(1).x() + f.vertex(2).x());
-    auto center_y = w / 2.f - (1.f / 3) * (f.vertex(0).y() + f.vertex(1).y() + f.vertex(2).y());
-    in[f] = (center_x * center_x + center_y * center_y > w / 3.) ? 1 : 0;
-  }
+  // for(auto& f : m.edges()) {
```
Comment (mroethlin) [inline dawn/prototype/driver.cpp]: delete commented out code?

File: dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp (hunk 2 above)
Comment (twicki) [reply dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp]: (this is your code :) )

**Pull request merged**